File: sdk/typescript/src/__tests__/realtime.test.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";

import { Torchwood } from "../graviton.js";
import type { RealtimeWebSocket } from "../client/realtime.js";

// MockWebSocket 模拟服务端一侧：记录客户端发出的帧，测试可手动触发
// open / 服务端帧 / close（如 token_expired）。
class MockWebSocket implements RealtimeWebSocket {
  static instances: MockWebSocket[] = [];

  onopen: (() => void) | null = null;
  onmessage: ((ev: { data: unknown }) => void) | null = null;
  onclose: ((ev: { code: number; reason: string }) => void) | null = null;
  onerror: (() => void) | null = null;

  readonly sent: string[] = [];
  closed = false;

  constructor(readonly url: string) {
    MockWebSocket.instances.push(this);
  }

  send(data: string): void {
    this.sent.push(data);
  }

  close(): void {
    this.closed = true;
  }

  frames(): Record<string, unknown>[] {
    return this.sent.map((s) => JSON.parse(s) as Record<string, unknown>);
  }

  serverOpen(): void {
    this.onopen?.();
  }

  serverSend(frame: Record<string, unknown>): void {
    this.onmessage?.({ data: JSON.stringify(frame) });
  }

  serverClose(code: number, reason: string): void {
    this.onclose?.({ code, reason });
  }
}

function factory(): (url: string) => RealtimeWebSocket {
  MockWebSocket.instances = [];
  return (url) => new MockWebSocket(url);
}

const tick = () => new Promise((r) => setTimeout(r, 0));
const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

describe("realtime", () => {
  it("hello 握手 + 订阅/事件/ping 帧序列", async () => {
    const client = Torchwood.withAccessToken("http://localhost:9080", "proj-1", "t1");
    const rt = client.realtime.connect({ webSocketFactory: factory() });
    await tick();
    const ws = MockWebSocket.instances[0];
    assert.equal(ws.url, "ws://localhost:9080/v1/realtime");

    ws.serverOpen();
    assert.deepEqual(ws.frames(), [
      { type: "hello", project_id: "proj-1", access_token: "t1" },
    ]);

    ws.serverSend({ type: "hello_ok", connection_id: "conn-1" });
    const events: unknown[] = [];
    const sub = rt.subscribe("databases.app.collections.posts", (ev) => events.push(ev));
    assert.deepEqual(ws.frames()[1], {
      type: "subscribe",
      id: "c1",
      channel: "databases.app.collections.posts",
    });

    ws.serverSend({ type: "subscribed", id: "c1", channel: "databases.app.collections.posts" });
    ws.serverSend({
      type: "event",
      channel: "databases.app.collections.posts",
      payload: { event_id: "e1", document_id: "d1" },
    });
    assert.deepEqual(events, [
      { channel: "databases.app.collections.posts", payload: { event_id: "e1", document_id: "d1" } },
    ]);

    // 服务端 30s ping → 客户端回 pong。
    ws.serverSend({ type: "ping" });
    assert.deepEqual(ws.frames().at(-1), { type: "pong" });

    // 退订：发送 unsubscribe 帧。
    sub.unsubscribe();
    assert.deepEqual(ws.frames().at(-1), {
      type: "unsubscribe",
      id: "c1",
      channel: "databases.app.collections.posts",
    });

    rt.close();
    assert.equal(rt.status, "closed");
  });

  it("Console cookie 场景：getAccessToken 省略时 hello 不带 access_token", async () => {
    const client = new Torchwood({ endpoint: "http://localhost:9080", projectId: "proj-1" });
    client.realtime.connect({ webSocketFactory: factory() });
    await tick();
    const ws = MockWebSocket.instances[0];
    ws.serverOpen();
    assert.deepEqual(ws.frames(), [{ type: "hello", project_id: "proj-1" }]);
  });

  it("token_expired 断线：刷新 token → 重新 hello → 重订全部频道（不补历史）", async () => {
    const client = Torchwood.withAccessToken("http://localhost:9080", "proj-1", "t1");
    const tokens = ["t2", "t3"];
    let refreshes = 0;
    const rt = client.realtime.connect({
      webSocketFactory: factory(),
      reconnectInitialDelayMs: 1,
      reconnectMaxDelayMs: 5,
      getAccessToken: () => (refreshes === 0 ? "t1" : tokens[Math.min(refreshes - 1, 1)]),
    });
    // 包一层计数（首次 connect 也会调 getAccessToken）。
    await tick();
    const ws1 = MockWebSocket.instances[0];
    ws1.serverOpen();
    ws1.serverSend({ type: "hello_ok", connection_id: "conn-1" });
    const events: unknown[] = [];
    rt.subscribe("databases.app.collections.posts", (ev) => events.push(ev));
    rt.subscribe("databases.app.collections.comments", () => {});
    assert.equal(ws1.frames().length, 3); // hello + 2 subscribe

    // JWT 到期：服务端以 StatusPolicyViolation + token_expired 关连接。
    refreshes += 1;
    ws1.serverClose(1008, "token_expired");
    assert.equal(rt.status, "reconnecting");
    await sleep(20);

    assert.equal(MockWebSocket.instances.length, 2);
    const ws2 = MockWebSocket.instances[1];
    ws2.serverOpen();
    // 重连 hello 带刷新后的 token。
    assert.deepEqual(ws2.frames(), [
      { type: "hello", project_id: "proj-1", access_token: "t2" },
    ]);
    ws2.serverSend({ type: "hello_ok", connection_id: "conn-2" });
    // 重订全部频道（复用原订阅 id），无历史补放。
    assert.deepEqual(ws2.frames().slice(1), [
      { type: "subscribe", id: "c1", channel: "databases.app.collections.posts" },
      { type: "subscribe", id: "c2", channel: "databases.app.collections.comments" },
    ]);
    assert.equal(rt.status, "connected");

    // 旧连接的迟到帧不再分发。
    ws1.serverSend({ type: "event", channel: "databases.app.collections.posts", payload: {} });
    ws2.serverSend({
      type: "event",
      channel: "databases.app.collections.posts",
      payload: { event_id: "e2" },
    });
    assert.deepEqual(events, [
      { channel: "databases.app.collections.posts", payload: { event_id: "e2" } },
    ]);

    rt.close();
  });

  it("指数退避：连续失败时延迟翻倍且封顶", async () => {
    const client = Torchwood.withAccessToken("http://localhost:9080", "proj-1", "t1");
    const rt = client.realtime.connect({
      webSocketFactory: factory(),
      reconnectInitialDelayMs: 5,
      reconnectMaxDelayMs: 20,
    });
    await tick();
    MockWebSocket.instances[0].serverOpen();
    MockWebSocket.instances[0].serverSend({ type: "hello_ok", connection_id: "c1" });

    // 第 1 次断线 → 5ms 后重连；第 2 次断线 → 10ms；第 3 次 → 20ms（封顶）。
    MockWebSocket.instances[0].serverClose(1006, "");
    await sleep(30);
    assert.equal(MockWebSocket.instances.length, 2);
    MockWebSocket.instances[1].serverClose(1006, "");
    await sleep(40);
    assert.equal(MockWebSocket.instances.length, 3);
    rt.close();

    // close 后不再重连。
    MockWebSocket.instances[2].serverClose(1006, "");
    await sleep(60);
    assert.equal(MockWebSocket.instances.length, 3);
    assert.equal(rt.status, "closed");
  });
});

File: sdk/go/client/realtime_test.go
package client

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"sync"
	"testing"
	"time"

	"github.com/coder/websocket"
	"github.com/stretchr/testify/require"
	clientv1 "github.com/torchwooddev/torchwood/genproto/client/v1"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// realtimeTestServer 是协议级假服务端（对齐 internal/api/realtime 的帧
// 格式）：记录每次 hello 的 token 与全部控制帧，暴露当前连接供测试
// 主动推事件 / 关连接。
type realtimeTestServer struct {
	mu         sync.Mutex
	subscribes []string

	url    string
	hellos chan string          // 每次握手的 access_token
	frames chan map[string]any  // 客户端控制帧（subscribe/unsubscribe/ping）
	conns  chan *websocket.Conn // 每次握手成功后的服务端连接
}

func newRealtimeTestServer(t *testing.T) *realtimeTestServer {
	t.Helper()
	s := &realtimeTestServer{
		hellos: make(chan string, 8),
		frames: make(chan map[string]any, 32),
		conns:  make(chan *websocket.Conn, 8),
	}
	ts := httptest.NewServer(http.HandlerFunc(s.serveWS))
	s.url = ts.URL
	t.Cleanup(ts.Close)
	return s
}

func (s *realtimeTestServer) serveWS(w http.ResponseWriter, r *http.Request) {
	conn, err := websocket.Accept(w, r, nil)
	if err != nil {
		return
	}
	defer conn.CloseNow()
	ctx := r.Context()

	_, data, err := conn.Read(ctx)
	if err != nil {
		return
	}
	var hello struct {
		Type        string `json:"type"`
		ProjectID   string `json:"project_id"`
		AccessToken string `json:"access_token"`
	}
	if err := json.Unmarshal(data, &hello); err != nil || hello.Type != "hello" || hello.ProjectID == "" {
		_ = writeServerFrame(conn, map[string]any{"type": "error", "code": "UNAUTHENTICATED", "message": "bad hello"})
		_ = conn.Close(websocket.StatusPolicyViolation, "UNAUTHENTICATED")
		return
	}
	s.hellos <- hello.AccessToken
	if err := writeServerFrame(conn, map[string]any{"type": "hello_ok", "connection_id": "conn-x"}); err != nil {
		return
	}
	s.conns <- conn

	for {
		_, data, err := conn.Read(ctx)
		if err != nil {
			return
		}
		var f map[string]any
		if err := json.Unmarshal(data, &f); err != nil {
			continue
		}
		s.frames <- f
		switch f["type"] {
		case "subscribe":
			s.mu.Lock()
			s.subscribes = append(s.subscribes, f["channel"].(string))
			s.mu.Unlock()
			_ = writeServerFrame(conn, map[string]any{"type": "subscribed", "id": f["id"], "channel": f["channel"]})
		case "ping":
			_ = writeServerFrame(conn, map[string]any{"type": "pong"})
		}
	}
}

func writeServerFrame(conn *websocket.Conn, f map[string]any) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	data, _ := json.Marshal(f)
	return conn.Write(ctx, websocket.MessageText, data)
}

func recvOf[T any](t *testing.T, ch <-chan T) T {
	t.Helper()
	select {
	case v := <-ch:
		return v
	case <-time.After(5 * time.Second):
		t.Fatal("等待服务端帧超时")
		var zero T
		return zero
	}
}

// realtimeTokens 初始 token（远期不过期，避免触发 refreshIfExpiring）。
func realtimeTokens() *clientv1.TokenBundle {
	return &clientv1.TokenBundle{
		AccessToken:  "jwt-1",
		RefreshToken: "refresh-1",
		ExpiresAt:    timestamppb.New(time.Unix(1893456000, 0)),
	}
}

func TestRealtime_HelloSubscribeEvent(t *testing.T) {
	srv := newRealtimeTestServer(t)
	c, _ := newTestClient(t, WithProjectID("proj-1"), WithInitialTokens(realtimeTokens()))

	rt, err := c.ConnectRealtime(context.Background(), srv.url)
	require.NoError(t, err)
	t.Cleanup(func() { _ = rt.Close() })

	require.Equal(t, "jwt-1", recvOf(t, srv.hellos))
	serverConn := recvOf(t, srv.conns)

	events := make(chan RealtimeEvent, 4)
	rt.Subscribe("databases.app.collections.posts", func(ev RealtimeEvent) { events <- ev })
	f := recvOf(t, srv.frames)
	require.Equal(t, "subscribe", f["type"])
	require.Equal(t, "databases.app.collections.posts", f["channel"])

	// 服务端推事件帧 → handler 收到（payload 无 acl）。
	require.NoError(t, writeServerFrame(serverConn, map[string]any{
		"type":    "event",
		"channel": "databases.app.collections.posts",
		"payload": map[string]any{"event_id": "e1", "document_id": "d1"},
	}))
	ev := recvOf(t, events)
	require.Equal(t, "databases.app.collections.posts", ev.Channel)
	require.Equal(t, "e1", ev.Payload["event_id"])

	// 服务端 ping → 客户端回 pong。
	require.NoError(t, writeServerFrame(serverConn, map[string]any{"type": "ping"}))
	f = recvOf(t, srv.frames)
	require.Equal(t, "pong", f["type"])
}

func TestRealtime_TokenExpiredReconnectResubscribes(t *testing.T) {
	srv := newRealtimeTestServer(t)
	c, fake := newTestClient(t, WithProjectID("proj-1"), WithInitialTokens(realtimeTokens()))
	// 刷新后的新 token（fakeAccount.RefreshToken 要求 refresh token 为 refresh-1）。
	fake.tokens = &clientv1.TokenBundle{
		AccessToken:  "jwt-2",
		RefreshToken: "refresh-1",
		ExpiresAt:    timestamppb.New(time.Unix(1893456000, 0)),
	}

	rt, err := c.ConnectRealtime(context.Background(), srv.url, WithRealtimeBackoff(time.Millisecond, 5*time.Millisecond))
	require.NoError(t, err)
	t.Cleanup(func() { _ = rt.Close() })

	require.Equal(t, "jwt-1", recvOf(t, srv.hellos))
	conn1 := recvOf(t, srv.conns)
	rt.Subscribe("databases.app.collections.posts", func(RealtimeEvent) {})
	rt.Subscribe("databases.app.collections.comments", func(RealtimeEvent) {})
	recvOf(t, srv.frames) // subscribe posts
	recvOf(t, srv.frames) // subscribe comments

	// JWT 到期：服务端以 StatusPolicyViolation + token_expired 关连接。
	require.NoError(t, conn1.Close(websocket.StatusPolicyViolation, "token_expired"))

	// 重连：先强制刷新（refreshCalls=1），hello 带新 token。
	require.Equal(t, "jwt-2", recvOf(t, srv.hellos))
	require.Equal(t, int32(1), fake.refreshCalls.Load())
	recvOf(t, srv.conns)

	// 重订全部频道（不补历史）。
	got := map[string]bool{}
	for i := 0; i < 2; i++ {
		f := recvOf(t, srv.frames)
		require.Equal(t, "subscribe", f["type"])
		got[f["channel"].(string)] = true
	}
	require.True(t, got["databases.app.collections.posts"])
	require.True(t, got["databases.app.collections.comments"])
}

func TestRealtime_HandshakeErrorSurfaces(t *testing.T) {
	srv := newRealtimeTestServer(t)
	// 无 ProjectID：直接报错。
	c, _ := newTestClient(t, WithInitialTokens(realtimeTokens()))
	_, err := c.ConnectRealtime(context.Background(), srv.url)
	require.Error(t, err)

	// hello 缺 project_id 的场景由服务端 error 帧覆盖（见 serveWS），这里
	// 断言坏 endpoint 报错。
	c2, _ := newTestClient(t, WithProjectID("proj-1"), WithInitialTokens(realtimeTokens()))
	_, err = c2.ConnectRealtime(context.Background(), "ftp://x")
	require.Error(t, err)
}

func TestRealtime_CloseStopsReconnect(t *testing.T) {
	srv := newRealtimeTestServer(t)
	c, _ := newTestClient(t, WithProjectID("proj-1"), WithInitialTokens(realtimeTokens()))

	rt, err := c.ConnectRealtime(context.Background(), srv.url, WithRealtimeBackoff(time.Millisecond, 5*time.Millisecond))
	require.NoError(t, err)
	recvOf(t, srv.hellos)
	recvOf(t, srv.conns)

	require.NoError(t, rt.Close())
	// 主动关闭后不再重连：hellos 通道应保持为空。
	select {
	case tok := <-srv.hellos:
		t.Fatalf("close 后不应重连，收到 hello token %q", tok)
	case <-time.After(100 * time.Millisecond):
	}
}
